src/radical/pilot/agent/executing/popen.py

```diff
@@ -499,9 +476,48 @@ def _check_running(self):
 
         return action
 
+    # --------------------------------------------------------------------------
+    #
+    def cancel_task(self, task):
+
+        if task not in self._tasks_to_watch:
+            # task completed meanwhile
+            return
+
+        tid = task['uid']
+
+        # FIXME: there is a race condition between the state poll
+        # above and the kill command below.  We probably should pull
+        # state after kill again?
+
+        self._prof.prof('task_run_cancel_start', uid=tid)
+
+        # We got a request to cancel this task - send SIGTERM to the
+        # process group (which should include the actual launch
+        # method)
+        task['proc'].kill()
+        try:
+            os.killpg(task['proc'].pid, signal.SIGTERM)
+        except OSError:
+            # task is already gone, we ignore this
+            pass
+        task['proc'].wait()  # make sure proc is collected
+
+        self._prof.prof('task_run_cancel_stop', uid=tid)
+
+        del task['proc']  # proc is not json serializable
+        self._prof.prof('unschedule_start', uid=tid)
+        self.publish(rpc.AGENT_UNSCHEDULE_PUBSUB, task)
+        self.advance(task, rps.CANCELED, publish=True, push=False)
+
+        # we don't need to watch canceled tasks
+        # FIXME: list action should be locked
+        self._tasks_to_watch.remove(task)
```

Comment:
should we have some "lock" attribute (`self._watch_lock     = mt.RLock()`?), which will be used when tasks are added to `_tasks_to_watch` and removed from it?